lib/metrics/context.js

```diff
@@ -6,43 +6,178 @@
+      events = [ events ]
+    }
+
+    if (! Array.isArray(events)) {
+      return P.resolve()
+    }
+
+    var memcached = getMemcached()
+
+    return P.all(events.map(function (event) {
+      return memcached.setAsync(getKey(token, event), metadata, config.memcached.lifetime)
+        .catch(function (err) {
+          log.error({ op: 'metricsContext.save', err: err })
+        })
+    }))
+  }
+
+  function getMemcached () {
+    try {
+      if (config.memcached.address === 'none') {
+        return NULL_MEMCACHED
       }
+
+      if (! _memcached) {
+        _memcached = new Memcached(config.memcached.address, {
+          timeout: 500,
+          retries: 1,
+          retry: 1000,
+          reconnect: 1000,
+          idle: config.memcached.idle,
+          namespace: 'fxa-metrics~'
+        })
+      }
+
+      return _memcached
+    } catch (err) {
+      log.error({ op: 'metricsContext.getMemcached', err: err })
+
+      return NULL_MEMCACHED
     }
+  }
+
+  /**
+   * Copies metrics context metadata to data, using either metadata
+   * passed in from a request or previously-saved metadata for a
+   * token.
+   *
+   * @param data    target object
+   * @param request request object
+   * @param event   event name
+   */
+  function copy (data, request, event) {
+    var metadata = request.payload && request.payload.metricsContext
+    var token = request.auth && request.auth.credentials
+    var doNotTrack = request.headers && request.headers.dnt === '1'
+
+    return P.resolve()
+      .then(function () {
+        if (metadata) {
+          return metadata
+        }
+
+        if (token && event) {
+          return restore(token, event)
+        }
+      })
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.copy', err: err })
+      })
+      .then(function (metadata) {
+        if (metadata) {
+          data.time = Date.now()
+          data.flow_id = metadata.flowId
+          data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
+          data.context = metadata.context
+          data.entrypoint = metadata.entrypoint
+          data.migration = metadata.migration
+          data.service = metadata.service
+
+          if (! doNotTrack) {
+            data.utm_campaign = metadata.utmCampaign
+            data.utm_content = metadata.utmContent
+            data.utm_medium = metadata.utmMedium
+            data.utm_source = metadata.utmSource
+            data.utm_term = metadata.utmTerm
+          }
+        }
 
-    return data
+        return data
+      })
   }
+
+  function restore (token, event) {
+    var memcached = getMemcached()
+    return memcached.getAsync(getKey(token, event))
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.restore', err: err })
+      })
+  }
+
+  /**
+   * Removes metrics context metadata for a token:event pair.
+   *
+   * @param token
+   * @param event
+   */
+  function remove (token, event) {
```

Comment:
You have many calls to this method that are immediately preceeded by logging an activity-event, and hence by a call to `metricsContext.restore`.  I wonder if it would simplify the logic to have `metricsContext.restore` automatically remove the item after it has been consumed, rather than having to remember to call `remove` in all the right places.